homeassistant/components/xiaomi_miio/humidifier.py

```diff
@@ -0,0 +1,660 @@
+    SUPPORT_MODES,
+)
+from homeassistant.const import ATTR_ENTITY_ID, CONF_HOST, CONF_NAME, CONF_TOKEN
+import homeassistant.helpers.config_validation as cv
+
+from .const import (
+    CONF_DEVICE,
+    CONF_FLOW_TYPE,
+    DOMAIN,
+    MODEL_AIRHUMIDIFIER_CA1,
+    MODEL_AIRHUMIDIFIER_CA4,
+    MODEL_AIRHUMIDIFIER_CB1,
+    MODELS_HUMIDIFIER,
+    MODELS_HUMIDIFIER_MIOT,
+    SERVICE_HUMIDIFIER_SET_BUZZER_OFF,
+    SERVICE_HUMIDIFIER_SET_BUZZER_ON,
+    SERVICE_HUMIDIFIER_SET_CHILD_LOCK_OFF,
+    SERVICE_HUMIDIFIER_SET_CHILD_LOCK_ON,
+    SERVICE_HUMIDIFIER_SET_DRY_OFF,
+    SERVICE_HUMIDIFIER_SET_DRY_ON,
+    SERVICE_HUMIDIFIER_SET_LED_BRIGHTNESS,
+    SERVICE_HUMIDIFIER_SET_LED_OFF,
+    SERVICE_HUMIDIFIER_SET_LED_ON,
+    SERVICE_HUMIDIFIER_SET_MOTOR_SPEED,
+    SERVICE_HUMIDIFIER_SET_TARGET_HUMIDITY,
+)
+from .device import XiaomiMiioEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "Xiaomi Miio Device"
+DEFAULT_MIN_HUMIDITY = 30
+DEFAULT_MAX_HUMIDITY = 80
+DATA_KEY = "humidifier.xiaomi_miio"
+
+CONF_MODEL = "model"
+
+ATTR_MODEL = "model"
+
+MODE_AUTO = "Auto"
+MODE_HIGH = "High"
+MODE_LOW = "Low"
+MODE_MEDIUM = "Medium"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_TOKEN): vol.All(cv.string, vol.Length(min=32, max=32)),
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_MODEL): vol.In(MODELS_HUMIDIFIER),
+    }
+)
+
+ATTR_TEMPERATURE = "temperature"
+ATTR_HUMIDITY = "humidity"
+ATTR_BUZZER = "buzzer"
+ATTR_CHILD_LOCK = "child_lock"
+ATTR_LED = "led"
+ATTR_LED_BRIGHTNESS = "led_brightness"
+ATTR_BRIGHTNESS = "brightness"
+ATTR_LEVEL = "level"
+ATTR_CURRENT_HUMIDITY = "current_humidity"
+ATTR_FEATURES = "features"
+ATTR_VOLUME = "volume"
+ATTR_USE_TIME = "use_time"
+ATTR_BUTTON_PRESSED = "button_pressed"
+
+# Air Humidifier
+ATTR_TARGET_HUMIDITY = "target_humidity"
+ATTR_TRANS_LEVEL = "trans_level"
+ATTR_HARDWARE_VERSION = "hardware_version"
+
+# Air Humidifier CA
+ATTR_MOTOR_SPEED = "motor_speed"
+ATTR_DEPTH = "depth"
+ATTR_DRY = "dry"
+
+# Air Humidifier CA4
+ATTR_ACTUAL_MOTOR_SPEED = "actual_speed"
+ATTR_FAHRENHEIT = "fahrenheit"
+ATTR_FAULT = "fault"
+ATTR_WATER_LEVEL = "water_level"
+
+AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_COMMON = {
+    ATTR_TEMPERATURE: "temperature",
+    ATTR_CURRENT_HUMIDITY: "humidity",
+    ATTR_BUZZER: "buzzer",
+    ATTR_CHILD_LOCK: "child_lock",
+    ATTR_LED_BRIGHTNESS: "led_brightness",
+    ATTR_USE_TIME: "use_time",
+}
+
+AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER = {
+    **AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_COMMON,
+    ATTR_TRANS_LEVEL: "trans_level",
+    ATTR_BUTTON_PRESSED: "button_pressed",
+    ATTR_HARDWARE_VERSION: "hardware_version",
+}
+
+AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_CA_AND_CB = {
+    **AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_COMMON,
+    ATTR_MOTOR_SPEED: "motor_speed",
+    ATTR_DEPTH: "depth",
+    ATTR_DRY: "dry",
+    ATTR_HARDWARE_VERSION: "hardware_version",
```

Comment:
Static info like hardware info should not be part of state attributes.

Reply:
Can be removed. I think nobody uses this attribute for automations. Is there a better place for static infos about the device?

Reply:
We have manufacturer, firmware version and model in device info. But it's not possible to add arbitrary info in device info yet.

https://developers.home-assistant.io/docs/device_registry_index#device-properties

Reply:
@syssi device info is added here:
https://github.com/home-assistant/core/blob/208a104e96290bcbc6760f62a4b4ba36c8cd5a86/homeassistant/components/xiaomi_miio/device.py#L78-L91
(That is general code for all miio devices)